From d46427f232011c7f65dcfcf2f90af8a3cbb2dd2d Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Wed, 29 Jun 2011 17:57:26 -0600
Subject: [PATCH] Add an apply button to the PatchScriptSettingsPanel

This allows a user to temporarily apply a setting without
having to alter their saved preferences.  All settings
are reset to the saved value whenenver the PatchScreen
is reloaded after an apply (unless not signed in). If
not signed in, the save button (renamed from update now)
will be invisible since the apply button will act the
same way an update used to for unsigned in users.

Change-Id: I6ee5d688fbcee268aef1de1408762904d2a3c1ea
---
 .../google/gerrit/client/patches/PatchScreen.java  |    3 ++
 .../client/patches/PatchScriptSettingsPanel.java   |   24 +++++++++----
 .../client/patches/PatchScriptSettingsPanel.ui.xml |   35 +++++++++++--------
 .../client/ui/ListenableAccountDiffPreference.java |   25 ++++++++++++++
 4 files changed, 65 insertions(+), 22 deletions(-)

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
index dcfb34a..32fef3d 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
@@ -148,6 +148,9 @@ public abstract class PatchScreen extends Screen implements
 
     prefs = fileList != null ? fileList.getPreferences() :
                                new ListenableAccountDiffPreference();
+    if (Gerrit.isSignedIn()) {
+      prefs.reset();
+    }
     prefs.addValueChangeHandler(
         new ValueChangeHandler<AccountDiffPreference>() {
           @Override
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.java
index bce7f55..b274656 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.java
@@ -92,6 +92,9 @@ public class PatchScriptSettingsPanel extends Composite implements
 
 
   @UiField
+  Button apply;
+
+  @UiField
   Button update;
 
   /**
@@ -115,6 +118,7 @@ public class PatchScriptSettingsPanel extends Composite implements
     initContext(context);
     if (!Gerrit.isSignedIn()) {
       reviewed.setVisible(false);
+      update.setVisible(false);
     }
 
     KeyPressHandler onEnter = new KeyPressHandler() {
@@ -220,7 +224,19 @@ public class PatchScriptSettingsPanel extends Composite implements
     update();
   }
 
+  @UiHandler("apply")
+  void onApply(ClickEvent event) {
+    apply();
+  }
+
   private void update() {
+    apply();
+    if (Gerrit.isSignedIn()) {
+      persistDiffPreferences();
+    }
+  }
+
+  private void apply() {
     if (colWidth.getIntValue() <= 0) {
       new ErrorDialog(PatchUtil.C.illegalNumberOfColumns()).center();
       return;
@@ -240,19 +256,13 @@ public class PatchScriptSettingsPanel extends Composite implements
     dp.setExpandAllComments(expandAllComments.getValue());
 
     listenablePrefs.set(dp);
-
-    if (Gerrit.isSignedIn()) {
-      persistDiffPreferences();
-    }
   }
 
   private void persistDiffPreferences() {
     setEnabled(false);
-    Util.ACCOUNT_SVC.changeDiffPreferences(getValue(),
-        new GerritCallback<VoidResult>() {
+    listenablePrefs.save(new GerritCallback<VoidResult>() {
       @Override
       public void onSuccess(VoidResult result) {
-        Gerrit.setAccountDiffPreference(getValue());
         setEnabled(true);
       }
 
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.ui.xml b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.ui.xml
index 8bda032..ce3140c 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.ui.xml
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScriptSettingsPanel.ui.xml
@@ -115,46 +115,51 @@ limitations under the License.
       </g:CheckBox>
     </td>
 
-    <td>
-      <g:CheckBox
-          ui:field='expandAllComments'
-          text='Expand All Comments'
-          tabIndex='9'>
-        <ui:attribute name='text'/>
-      </g:CheckBox>
-    </td>
-
     <td rowspan='2'>
       <g:CheckBox
           ui:field='skipUncommented'
           text='Skip Uncommented Files'
-          tabIndex='10'>
+          tabIndex='9'>
         <ui:attribute name='text'/>
       </g:CheckBox>
       <br/>
       <g:CheckBox
           ui:field='skipDeleted'
           text='Skip Deleted Files'
-          tabIndex='11'>
+          tabIndex='10'>
         <ui:attribute name='text'/>
       </g:CheckBox>
     </td>
 
-    <td valign='bottom' rowspan='2'>
+    <td rowspan='2'>
+      <g:CheckBox
+          ui:field='expandAllComments'
+          text='Expand All Comments'
+          tabIndex='11'>
+        <ui:attribute name='text'/>
+      </g:CheckBox>
+      <br/>
       <g:Button
-          ui:field='update'
-          text='Update'
+          ui:field='apply'
+          text='Apply'
           styleName='{style.updateButton}'
           tabIndex='12'>
         <ui:attribute name='text'/>
       </g:Button>
+      <g:Button
+          ui:field='update'
+          text='Save'
+          styleName='{style.updateButton}'
+          tabIndex='13'>
+        <ui:attribute name='text'/>
+      </g:Button>
     </td>
 
     <td>
       <g:CheckBox
           ui:field='reviewed'
           text='Reviewed'
-          tabIndex='13'>
+          tabIndex='14'>
         <ui:attribute name='text'/>
       </g:CheckBox>
     </td>
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableAccountDiffPreference.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableAccountDiffPreference.java
index 195adf5..41de2cd 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableAccountDiffPreference.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/ListenableAccountDiffPreference.java
@@ -15,12 +15,37 @@
 package com.google.gerrit.client.ui;
 
 import com.google.gerrit.client.Gerrit;
+import com.google.gerrit.client.account.Util;
+import com.google.gerrit.client.rpc.GerritCallback;
 import com.google.gerrit.reviewdb.AccountDiffPreference;
+import com.google.gwtjsonrpc.client.VoidResult;
 
 public class ListenableAccountDiffPreference
     extends ListenableValue<AccountDiffPreference> {
 
   public ListenableAccountDiffPreference() {
+    reset();
+  }
+
+  public void save(final GerritCallback<VoidResult> cb) {
+    if (Gerrit.isSignedIn()) {
+      Util.ACCOUNT_SVC.changeDiffPreferences(get(),
+          new GerritCallback<VoidResult>() {
+        @Override
+        public void onSuccess(VoidResult result) {
+          Gerrit.setAccountDiffPreference(get());
+          cb.onSuccess(result);
+        }
+
+        @Override
+        public void onFailure(Throwable caught) {
+          cb.onFailure(caught);
+        }
+      });
+    }
+  }
+
+  public void reset() {
     if (Gerrit.isSignedIn() && Gerrit.getAccountDiffPreference() != null) {
       set(Gerrit.getAccountDiffPreference());
     } else {
-- 
1.7.3.1

